src/librustc_resolve/build_reduced_graph.rs

```diff
@@ -414,7 +425,18 @@ impl<'a> Resolver<'a> {
         // value namespace, they are reserved for possible future use.
         let ctor_kind = CtorKind::from_ast(&variant.node.data);
         let ctor_def = Def::VariantCtor(def_id, ctor_kind);
-        self.define(parent, ident, ValueNS, (ctor_def, vis, variant.span, expansion));
+        let mut ctor_vis = vis;
+
+        let has_non_exhaustive = variant.node.attrs.iter()
+            .any(|item| item.check_name("non_exhaustive"));
+
+        // If the variant is marked as non_exhaustive then lower the visibility
+        // of constructor within the crate.
+        if has_non_exhaustive && ctor_vis == ty::Visibility::Public {
+            ctor_vis = ty::Visibility::Restricted(DefId::local(CRATE_DEF_INDEX));
+        }
+
+        self.define(parent, ident, ValueNS, (ctor_def, ctor_vis, variant.span, expansion));
```

Comment:
Since we don't support variants yet, this change can be removed.

Reply:
Resolved.